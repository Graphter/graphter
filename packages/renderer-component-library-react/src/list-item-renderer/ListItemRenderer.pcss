@import "graphter-theme/src/colours.pcss";

.title{
  font-size: 2rem;
}
.subtext{
  display: block;
  color: var(--dark-subtext)
}