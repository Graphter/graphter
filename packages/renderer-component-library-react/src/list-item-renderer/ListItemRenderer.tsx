import React, { Fragment } from 'react'
import * as R from 'ramda';
import s from './ListItemRenderer.pcss';
import { RenderItemProps } from "@graphter/renderer-react";

export interface ListItemRendererProps extends RenderItemProps {
  idPath?: Array<string>,
  titlePath?: Array<string>,
  subtextPath?: Array<string>
}

export default function ListItemRenderer({  idPath, item, titlePath, subtextPath }: ListItemRendererProps){

  const id: any = idPath ?
    R.path(idPath, item) :
    getFirstWithValue([['id'], ['Id'], ['ID'], ['key'], ['Key']], item);

  const title: any = titlePath ?
    R.path(titlePath, item) :
    getFirstWithValue([['name'], ['Name'], ['title'], ['Title'], ['id'], ['Id'], ['ID']], item);

  const subtext: any = subtextPath ?
    R.path(subtextPath, item) :
    getFirstWithValue([['description'], ['Description']], item);

  return (
    <Fragment>
      <span className={s.title} data-testid='title'>{title}</span>
      {subtext && <span className={s.subtext} data-testid='subtext'>{subtext}</span>}
    </Fragment>
  );
}

function getFirstWithValue(paths: Array<Array<string>>, obj: object){
  let value = null;
  paths.some(path => {
    value = R.path(path, obj);
    return value;
  });
  return value;
}