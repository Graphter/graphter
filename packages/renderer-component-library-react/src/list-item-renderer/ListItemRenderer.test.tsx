import React from "react";
import ListItemRenderer from "./ListItemRenderer";
import { act, render, within } from '@testing-library/react';
import { screen } from '@testing-library/dom'
import '@testing-library/jest-dom/extend-expect'
import { shallow } from 'enzyme';

describe('<ListItem />', () => {

  describe(`when no paths are passed in`, () => {

    it.each([ 'name', 'Name', 'title', 'Title', 'id', 'Id', 'ID' ])(
      `should look for a property called %p to guess as title`, async (prop) => {
        const { container, getByText } = render(<ListItemRenderer item={{
          somethingElse: 'blah blah',
          [prop]: 'Some name',
          description: 'Some descriptive text'
        }} />);
        expect(getByText('Some name')).not.toBeNull();
        expect(container).toMatchSnapshot();
      });

    it.each([ 'description', 'Description' ])(
      `should look for a property called %p to guess as subtext`, async (prop) => {
        const { container, getByText } = render(<ListItemRenderer item={{
          somethingElse: 'blah blah',
          name: 'Some Name',
          [prop]: 'Some descriptive text'
        }}/>);
        expect(getByText('Some descriptive text')).not.toBeNull();
        expect(container).toMatchSnapshot();
      });

  });

  it('should use titlePath and subtextPath when defined', async () => {
    const { getByText } = render(<ListItemRenderer
      titlePath={[ 'aSpecialName' ]}
      subtextPath={[ 'aSpecialDescription' ]}
      item={{
        aSpecialName: 'Some Special Name',
        aSpecialDescription: 'blah blah special blah',
      }}
    />);
    expect(getByText('Some Special Name')).not.toBeNull();
    expect(getByText('blah blah special blah')).not.toBeNull();
  });
});

export {};