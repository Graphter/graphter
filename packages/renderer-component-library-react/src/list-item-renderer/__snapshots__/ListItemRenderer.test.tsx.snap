// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListItem /> when no paths are passed in should look for a property called "Description" to guess as subtext 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some Name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "ID" to guess as title 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "Id" to guess as title 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "Name" to guess as title 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "Title" to guess as title 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "description" to guess as subtext 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some Name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "id" to guess as title 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "name" to guess as title 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;

exports[`<ListItem /> when no paths are passed in should look for a property called "title" to guess as title 1`] = `
<div>
  <span
    class="title"
    data-testid="title"
  >
    Some name
  </span>
  <span
    class="subtext"
    data-testid="subtext"
  >
    Some descriptive text
  </span>
</div>
`;
